### PR TITLE
feat: support multiple provisioning profiles

### LINKED ISSUE #953: Wrongly created exportOptions.plist file - multiple provisioning profile support
# Bug Report
When there is more than one target and you are trying to sign both it doesn't work right. 
This is partially a bug and partially a feature request.

### What is expected to happen?
I expect to be able to sign two or more targets from CI server

### What does actually happen?
I can't do it right now.

## Information
This is after reviewing the discussion here:
#820 and #538 
and #545 (which I think tried to solve this as well).

### Command or Code
I'm building an ios cordova app with a share extension.
I configured a provisioning profile for both the extension and the app
I setup everything up and the build succeeds but the export fails.
I took a look in the exportOptions.plist file and there's only one provisioning profile, somehow the bundle name and provision profile got mixed (bundle id is for the app extension and PP UUID is for the app) which I can't really explain, but in any case it doesn't support multiple PP which is what needed in order to be able to sign the app from a CI server.
Code can be seen here (although it's not straight forward)
https://github.com/IsraelHikingMap/Site/tree/bc641a1b8c6e22293678ebe8e1a214fa2782ef05/IsraelHiking.Web
Relevant build:
https://ci.appveyor.com/project/IsraelHikingHost/site/builds/34309662

### Environment, Platform, Device
iOS - Appveyor CI - macOS VM.

### Version information
cordova cli 9.0.0
cordova-ios: 6.1.0
macOS VM from appveyor and also my own: 10.15.5

## Checklist
- [x] I searched for existing GitHub issues
- [x] I updated all Cordova tooling to most recent version
- [x] I included all the necessary information above

Let me know if this is something on the road map or in implementation, otherwise I can try and see if I can do a PR for this...

### PR DESCRIPTION
**Pull request by HarelM**
### Platforms affected
ios

### Motivation and Context
When using a plugin that creates a new target it is currently impossible to sign it using cordova command line.
See issue #953 

Resolves #953

### Description

Added the option to set multiple provisioning profiles only in the build.json file.

### Testing
I've tested this code on the app I'm writing.

### Checklist

- [x] I've run the tests to see all new and existing tests pass
~I added automated test coverage as appropriate for this change~ Couldn't understand how... :-(
- [x] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
~I've updated the documentation if necessary~ not sure how to do it since this is only available when using a json file.

### CODE REVIEW

**Comment by HarelM** [comment]
@maintainers anyone?

**Comment by timbru31** [comment]
Can you name the issues this PR would close? I see some references but not sure if any of those should be closed with this PR?

File: bin/templates/scripts/cordova/lib/build.js
Comment (timbru31) [inline bin/templates/scripts/cordova/lib/build.js:242]: remove or use `events.emit` like other log statements.

**Review by timbru31 — CHANGES_REQUESTED** [review]

**Comment by HarelM** [comment]
@timbru31 Updated the name, hope this is what you intended.

File: bin/templates/scripts/cordova/lib/build.js
Comment (HarelM) [reply bin/templates/scripts/cordova/lib/build.js:242]: No problem, I'll check how other places do logging.

**Comment by HarelM** [comment]
@timbru31 done! Thanks for the review.
Let me know when a new version is available. :-)

**Comment by HarelM** [comment]
@timbru31 anything else that I need to do? Can this be merged?

**Comment by HarelM** [comment]
Any one? Can this be merged? I would really like to migrate of my fork and to this repo, hopefully using a npm version, but a git commit will also do as long as I'm not using my getting-out-dated-fork... :-)

**Comment by HarelM** [comment]
Anything I can do in order to move this request forward?

**Comment by brodybits** [comment]
I would recommend that you follow up with us on the mailing list or on Slack.

**Comment by HarelM** [comment]
@brodybits I tried slack and didn't got any response to my post...
I'm currently using this my fork in a production app which behaves as expected, so I'm guessing this is as good as it gets in terms of testing. And in general I don't think this change is a breaking change as I wrote it with backwards compatibility in mind...
I really hoped it wouldn't take so much time to merge a pull request... :-(

**Comment by jlews** [comment]
I'm currently really blocked by this. Does anyone know of a workaround that will allow me to edit the `exportOptions.plist` file after cordova generates it, but before the archive is exported? From what I can tell, this all happens inside the `cordova compile` command and there's no hook that will allow me to do this.

**Comment by HarelM** [comment]
You can fork cordova-ios and apply this change if you want the latest version or you can take my fork and use it in `package.json` as a git reference and not use the official vesion.

**Comment by CodeWithOz** [comment]
@HarelM can you confirm that the multiple profile usage of `build.json` suggested by @ollm works with your fork? I've tried using it for my multiple profiles (main app and share extension) but the build step fails with the error `error: "ShareExt" requires a provisioning profile with the App Groups feature. Select a provisioning profile in the Signing & Capabilities editor. (in target 'ShareExt' from project '{Project Name}')`. Of course the `ShareExt` provisioning profile actually exists and is specified in the `build.json` using as @ollm suggested, so that means the build step is not finding it correctly.

**Comment by CodeWithOz** [comment]
Okay thanks for confirming. Can you share the alternate solution you found please? And on my end yes I've confirmed that the profile has the App Groups feature.

**Comment by ollm** [comment]
Yes, I still use them to display images in push notifications

First I recommend that you check if the two or more profiles appear in the file `./projectName/platforms/ios/exportOptions.plist` after build, I build using the command line like this `cordova build ios --device --realese --buildConfig build.json`.

Like this:
``` xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>compileBitcode</key>
    <false/>
    <key>method</key>
    <string>app-store</string>
    <key>iCloudContainerEnvironment</key>
    <string>Production</string>
    <key>teamID</key>
    <string>XXXXXX</string>
    <key>provisioningProfiles</key>
    <dict>
      <key>app.package.name</key>
      <string>xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx</string>
      <key>app.package.name.NotificationService</key>
      <string>xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx</string>
    </dict>
    <key>signingStyle</key>
    <string>manual</string>
    <key>signingCertificate</key>
    <string>iPhone Distribution</string>
  </dict>
</plist>
```

If the profiles don't appear, open the files `./projectName/platforms/ios/cordova/lib/build.js` and `./projectName/platforms/ios/cordova/build` and check that the @HarelM changes are present, if they are not you can edit these same files manually or reinstall the `cordova-ios` fork.

**Comment by HarelM** [comment]
I basically associate the files and urls with my app, nothing else.
I use some plugins for that, mainly web intent and deep links.
You can see my code in the following places:
iOS hack:
https://github.com/IsraelHikingMap/Site/blob/be555a25e37b6d274d11f74a998986a5b27881a1/IsraelHiking.Web/src/index.html#L137
The main service I use:
https://github.com/IsraelHikingMap/Site/blob/be555a25e37b6d274d11f74a998986a5b27881a1/IsraelHiking.Web/src/application/services/open-with.service.ts
And a lot of definitions in the config.xml:
https://github.com/IsraelHikingMap/Site/blob/be555a25e37b6d274d11f74a998986a5b27881a1/IsraelHiking.Web/config.xml

**Comment by CodeWithOz** [comment]
@ollm thanks for the additional info, I'll check to see if that fixes my problem.

@HarelM thanks a lot for sharing. I don't use angular so please confirm if this is a correct description of your workaround: you use the `config.xml` file to specify all the file types that your app can receive from a share, and when the share happens (in the form of a deep link), you parse the shared information and route to the appropriate part of your app. Is that accurate?

If it is, I have one more question: did you have to take any additional steps to make your app a share target for those file types, or was it sufficient to only specify them in the `config.xml` file?

**Comment by ollm** [comment]
Something that I have forgotten to comment on is that I think it is necessary to have `Automatically manage signing` disabled (That's how I have it), although I'm not entirely sure.

Profile

**Comment by HarelM** [comment]
@CodeWithOz in general, yes.
The index.html is not related to angular and is super important for iOS.
I use other plugins to handle intent in android, I don't remember now exactly which one, you can check my package.json file...

**Comment by SteveGilvarry** [comment]
Can I ask what is preventing merging this and what I can do to assist to move it forward? Would be very useful for my notification service extension allowing me to build via command line, maybe not everyones use case but it works and doesn't impact those not using them.

**Comment by HarelM** [comment]
BTW, I'm currently migrating to Capacitor where this is not longer a concern as you simply create the `exportOptions.plist` file, check it in and use it how ever you see fit.
I did returned to this code in order to understand how to write the file and the command line to archive and export in xcode, but other than that this is no longer needed :-)

**Comment by Makishima** [comment]
Could anyone say if there any workaround to use several profiles with cordova build?

**Comment by HarelM** [comment]
This was my workaround, it didn't got merged, so I don't think there's another option besides building this manually from xcode...

**Review by erisu — APPROVED** [review]
Code LGTM

Tested the following commands to confirm that the standard process continues to build successfully. (with & without `build.json` configurations)

- [x] `cordova build ios`
- [x] `cordova build ios --debug`
- [x] `cordova build ios --release`
- [x] `cordova run ios` (simulator) (ad-hoc)

I also uploaded the built release (ad-hoc) ipa to a physical device running iOS 15.6.1.

As for testing multiple provisioning profiles, I do not have a project to confirm the feature/settings.

**Comment by erisu** [comment]
Can you rebase against the main branch?

The PR has changes to [`bin/templates/scripts/cordova/build`](https://github.com/apache/cordova-ios/pull/956/files#diff-6c53c92102136f10ec331ad99c14d0caf3631038ca90e1925e8e628ccad6d542) which has been removed.

**Comment by HarelM** [comment]
This PR is two years old now. I can't say I fully remember what I did there although the change wasn't big, so in theory I should be able to remember.
Having said that, I no longer use Cordova as I migrated to Capacitor, so feel free to copy the code and use it, or create a different PR.
Sorry...

**Comment by ollm** [comment]
I have send a pull request with the changes on the master branch https://github.com/apache/cordova-ios/pull/1251, I can confirm that these changes work on `cordova-ios` version `6.2.0`, but I haven't tested it on the master version yet.

Could you tell me how can I install `cordova-ios` master version in my project to test it?

**Comment by ollm** [comment]
> @ollm any of these variants should work.
> 
> If you want to install the main branch of your fork, on github.
> 
> * `cordova platform add github:ollm/cordova-ios`
> * `cordova platform add git@github.com:ollm/cordova-ios.git`
> * `cordova platform add https://github.com/ollm/cordova-ios.git`
> 
> _Note: The path contains `ollm`, which represents your fork of the repo. If you want to use anyone elses fork, just update the path._
> 
> Or you can install from a local directory.
> 
> * `cordova platform add /path/to/cordova-ios`

Thanks, I have tested the changes and it works perfectly with multiple provisioning profiles (I have tested with 2 profiles)

If there is anything else I need to do to get the branch merged, please let me know (eg documentation)

**Pull request closed without merging**
